build.gradle

```diff
@@ -1521,6 +1522,54 @@ project(':group-coordinator') {
   srcJar.dependsOn 'processMessages'
 }
 
+project(':test-common') {
+  base {
+    archivesName = "kafka-test-common"
+  }
+
+  configurations {
+    generator
+  }
+
+  dependencies {
+    testImplementation libs.scalaJava8Compat
```

Comment:
@mumrah `test-common` only depends on production code, so each module's test code can depend on `test-common` without creating circular dependencies. Furthermore, this approach eliminates the need for using SPI

Reply:
Thanks @chia7712, this is a nice improvement. I'm happy to see this code move out of core :) 

The reasoning for using the runtime classpath + SPI approach is to completely break the compile-time dependency on core for both test and main sources. 

What I had prototyped was:

* `:test-common` include all the common test implementation code (including dependency on `:core` for BrokerServer, ControllerServer, etc) -- this is what you've done in this PR
* `:test-common:api` new sub-module that just includes interfaces including our junit extension

Then a module (like `:metadata` for example) would have the following:

```
dependencies {
  // regular compile deps
  testImplementation project(":test-common:api")
  testRuntimeOnly project (":test-common") 
}
```

Does that make sense?

Reply:
@mumrah This PR is intended to demonstrate an alternative approach and identify potential blockers (such as [KAFKA-17659](https://issues.apache.org/jira/browse/KAFKA-17659)). It's perfectly fine to close this PR if a better approach, such as SPI, is adopted. Hence, please feel free to correct my if I misunderstand your prototype.

> The reasoning for using the runtime classpath + SPI approach is to completely break the compile-time dependency on core for both test and main sources.

The isolation of compile-time is good, but what the benefit we can get from it except for breaking circular dependencies?

> Then a module (like :metadata for example) would have the following:

not sure whether all modules can not reference to common test implementation code. For example, `connect` and `streams` will use `KafkaClusterTestKit` to build their embedded systems. Hence, their dependencies looks like this:
```
dependencies {
  
  // testImplementation project(":test-common:api") they don't use junit extension
  testImplementation project (":test-common") 
}
```

Reply:
> The isolation of compile-time is good, but what the benefit we can get from it except for breaking circular dependencies?

Longer term, decoupling the module's compile-time dependencies will help with Gradle incremental build and test avoidance. 

> For example, connect and streams will use KafkaClusterTestKit to build their embedded systems.

Maybe we need to figure out why those systems are using KafkaClusterTestKit instead of `@ClusterTest` and add missing features to our junit extensions.

Reply:
> Longer term, decoupling the module's compile-time dependencies will help with Gradle incremental build and test avoidance.

Yes, that is a significant benefit, but we can (almost) achieve those features without relying on SPI if we isolate the core-related code in test-common module, right?


> Maybe we need to figure out why those systems are using KafkaClusterTestKit instead of @ClusterTest and add missing features to our junit extensions.

there was a discussion (https://issues.apache.org/jira/browse/KAFKA-17174). I will start another discussion after this isolation get completed.

>  I wonder if we should put the code in src/main/java instead of src/test/java.

sure

Reply:
> but we can (almost) achieve those features without relying on SPI if we isolate the core-related code in test-common module, right?

Ok, it took me a while, but I finally understand what's happening. It's the explicit test dependencies in ":core" that are causing circularity. 

E.g., if metadata test depends on core tests to gain `@ClusterTest`, we get a circular dependency because the core tests actually depend on the metadata test output.

By creating "test-common" and "test-common-api" and having them _not_ depend on other module's test code, then we do break the circularity. 

--- 

Regarding incremental builds and test avoidance, I was surprised to find that the tests do not re-run when I modify something in core. I would think that the transitive dependency from metadata:test -> test-common-api -> core would cause a rebuild. Maybe this has to do with `implementation` vs `api` dependency configurations?

Reply:
> Regarding incremental builds and test avoidance, I was surprised to find that the tests do not re-run when I modify something in core. I would think that the transitive dependency from metadata:test -> test-common-api -> core would cause a rebuild. Maybe this has to do with implementation vs api dependency configurations?

That is optimized by gradle [Compilation Avoidance](https://docs.gradle.org/current/userguide/caching_java_projects.html?_gl=1*18t1ppj*_gcl_au*MTU2NjE0OTE0MC4xNzIxNjUwNzMz*_ga*MTU1MDI3MTQ5Mi4xNzIxNjUwNzMz*_ga_7W7NC6YNPT*MTcyNzgzNDg4OS4yOS4wLjE3Mjc4MzQ4ODkuNjAuMC4w#java_compilation). to quote:
> Since the cache key is only influenced by the ABI of dependencies (and not by their implementation details like private types and method bodies), task output caching can also reuse compiled classes if they were produced by the same sources and ABI-equivalent dependencies.

By reducing dependencies, ABI incompatibility (due to changes in `core`) is not propagated to the `metadata` module. Changes in the `core` module trigger the rebuild of `test-common` and `test-common-api` because of the implementation dependency, but the metadata module does not rebuild since it has no direct dependency on the `core` module.